executor/src/witgen/data_structures/caller_data.rs

```diff
@@ -0,0 +1,73 @@
+use itertools::Itertools;
+use powdr_number::FieldElement;
+
+use crate::witgen::{
+    machines::LookupCell,
+    processor::{Left, OuterQuery},
+    EvalError, EvalResult, EvalValue,
+};
+
+/// A representation of the caller's data.
+///
+/// Useful for implementing [Machine::process_plookup] in terms of [Machine::process_lookup_direct].
+pub struct CallerData<'a, 'b, T> {
+    /// The raw data of the caller. Unknown values should be ignored.
+    data: Vec<T>,
+    /// The affine expressions of the caller.
+    left: &'b Left<'a, T>,
+}
+
+impl<'a, 'b, T: FieldElement> From<&'b OuterQuery<'a, '_, T>> for CallerData<'a, 'b, T> {
+    /// Builds a `CallerData` from an `OuterQuery`.
+    fn from(outer_query: &'b OuterQuery<'a, '_, T>) -> Self {
+        let data = outer_query
+            .left
+            .iter()
+            .map(|l| l.constant_value().unwrap_or_default())
+            .collect();
+        Self {
+            data,
+            left: &outer_query.left,
+        }
+    }
+}
+
+impl<'a, 'b, T: FieldElement> CallerData<'a, 'b, T> {
+    /// Returns the data as a list of `LookupCell`s, as expected by `Machine::process_lookup_direct`.
+    pub fn as_lookup_cells(&mut self) -> Vec<LookupCell<'_, T>> {
+        self.data
+            .iter_mut()
+            .zip_eq(self.left.iter())
+            .map(|(value, left)| match left.constant_value().is_some() {
+                true => LookupCell::Input(value),
+                false => LookupCell::Output(value),
+            })
+            .collect()
+    }
+}
+
+impl<'a, 'b, T: FieldElement> From<CallerData<'a, 'b, T>> for EvalResult<'a, T> {
+    /// Turns the result of a direct lookup into an `EvalResult`, as used by `Machine::process_plookup`.
+    ///
+    /// Note that this function assumes that the lookup was successful and complete.
+    fn from(data: CallerData<'a, 'b, T>) -> EvalResult<'a, T> {
+        let mut result = EvalValue::complete(vec![]);
+        for (l, v) in data.left.iter().zip_eq(data.data.iter()) {
+            if !l.is_constant() {
```

Comment:
Maybe it would be better to store that? Essentially all three function call `is_constant` or `constant_value` on the items of `data.left`.

Reply:
Hmm, but should be fine, no? Do you expect it to be expensive? Their implementation is pretty straight-forward.

I'd like to not store redundant information if possible.